tools/mkoccdeps/mkoccdeps.1
.\" mkoccdeps command man page, last updated 30/12/2005 by Fred Barnes
.TH MKOCCDEPS "1" "December 2005" "kroc 1.4.0" KRoC
.SH NAME
mkoccdeps \- generates occam program dependency information
.SH SYNOPSIS
.B mkoccdeps
[\fIoption\fR [\fI...\fR]] \fIfile\fR...
.SH DESCRIPTION
.PP
Calculates dependencies for occam programs.  The programs scans the files given on the
command-line (as \fIfile\fR...) and produces Makefile compatible dependencies for them.
.PP
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
print usage summary and options
.TP
\fB\-V\fR, \fB\-\-version\fR
print version and exit
.TP
\fB\-v\fR, \fB\-\-verbose\fR
verbose operation
.TP
\fB\-o\fR, \fB\-\-output\fR  \fIFILE\fR
output to \fIFILE\fR.  "-" means standard output, that is the default
.TP
\fB\-f\fR, \fB\-\-follow\fR
follow referenced files and generate dependency information for those too
.TP
\fB\-F\fR, \fB\-\-follow\-local\fR
follow referenced files, but only those in the current directory
.TP
\fB\-a\fR, \fB\-\-allobjs\fR
additionally generate an ALLOBJS variable in the output containing all the object
files that have corresponding source files in the project
.TP
\fB\-A\fR, \fB\-\-allotherobjs\fR
generate an ALLOBJS variable in the output contaning all the object files that
have corresponding source files, except those whose source file was given on the
command-line (only makes sense with the \fB\-\-follow\fR option).
.TP
\fB\-I\fR, \fB\-\-include\fR  \fIPATH\fR
search \fIPATH\fR for occam libraries (sources) and includes
.PP
.SH EXAMPLE
.PP
For a typical project, containing multiple source files and where "main.occ" has the
top-level process, depdencies suitable for a Makefile can be generated using:
.PP
	mkoccdeps -o DEPS -F -A main.occ
.PP
This will examine main.occ and any files it references, and so on.  An ALLOBJS variable
will be output that contains all the object files except "main.o".  This can be used
in conjunction with a Makefile rule of the form:
.PP
	main:    $(ALLOBJS) main.occ
.br
	         kroc main.occ $(ALLOBJS) $(LIBS)
.PP
.SH DEFICIENCIES
.PP
The program is blind to any conditional compilation (using the pre-processor)
.PP
.SH AUTHOR
This manual page is maintained by Fred Barnes <frmb@kent.ac.uk>
.SH "SEE ALSO"
kroc(1)
.SH "REPORTING BUGS"
Please report bugs to <kroc-bugs@kent.ac.uk>
.SH COPYRIGHT
Copyright \(co 2004-2005 Fred Barnes <frmb@kent.ac.uk>
.PP
This program is free software; you can redistribute it
and/or modify it under the terms of the GNU General Public
License as published by the Free Software Foundation;
either version 2 of the License, or (at your option) any
later version.
.PP
This program is distributed in the hope that it will be
useful, but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU General Public License for more
details.
.PP
You should have received a copy of the GNU General Public
License along with this program; if not, write to the Free
Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
02139, USA.



CHANGELOG

KRoC/Linux Change-log (most recent first)
=========================================

26/03/2008	frmb	version up to 1.5.0-pre4
17/11/2007	cgr	fixes for some mobile process issues
16/11/2007	cgr	various CCSP updates
16/11/2007	ats	various updates to sdlraster, raster and occade
14/11/2007	frmb	tagged and released 1.5.0-pre3
14/11/2007	frmb	version-up to 1.5.0-pre3, added some sdlraster examples (using miniraster)
14/11/2007	frmb	restructured source packages into a single tree
13/11/2007	cgr	various updates, merged MP branch into trunk
31/10/2007	frmb	put 1.5.0-pre2 together (MP version, auto-detected)
31/10/2007	frmb	various build script updates
28/10/2007	frmb	version up to 1.5.0-pre2, added David Wood's robot programs
28/10/2007	frmb	small fix for remembering UDC build option
28/10/2007	frmb	moved source code repository to CSProjects, http://projects.cs.kent.ac.uk/
23/09/2007	frmb	version-up to 1.5.0 pre-releases
21/09/2007	ats	restructured distribution tree
21/09/2007	cgr	various updates
25/07/2007	ats	various restructuring of the source tree
14/07/2007	cgr	some work towards darwin support (for Mac OS X on Intel)
12/07/2007	frmb	put 1.4.1-pre6 together
11/07/2007	ats	added occSDL library and occade to the distribution
19/06/2007	ats	various package updates for module mechanism
19/06/2007	ats	updates to KRoC wrapper script
18/06/2007	ats	added -C option to tranx86 to dump #PRAGMA COMMENTs
17/06/2007	frmb	minor updates for NOCC
14/04/2007	cgr	CCSP fixes for RMoX
04/04/2007	ats	occam-doc updates and mechanical occam-doc-ifying of inmos libraries
06/03/2007	frmb	added NOCC path to build configuration (passed to kroc script configure)
06/03/2007	frmb	various tranx86 updates
14/02/2007	ats	occamdoc updates
03/02/2007	phw	updates to course sort-pump examples
19/01/2007	frmb	fixed a small bug in occ21 (ack to db99 for spotting it)
18/01/2007	ats	occamdoc updates
28/12/2006	frmb	added MoinMoin Wiki occam-pi parser to the distribution (in doc/)
20/12/2006	frmb	added occam-pi adventure game to the distribution
16/12/2006	frmb	put 1.4.1-pre5 together
01/12/2006	frmb	various updates to occamdoc and some packages
08/09/2006	frmb	partial updates to support building on 64-bit AMD Opteron / Linux machines
02/09/2006	frmb	various updates to tranx86 and CCSP
20/08/2006	frmb	added MySQL occam API (moa) to the contrib area
15/05/2006	ats	occamdoc moved within repository, build script updated
15/05/2006	frmb	put 1.4.1-pre4 together
15/05/2006	frmb	merged in Adam's occamdoc tool
14/05/2006	frmb	fixed some of the interfacing code in CCSP, normalised a bit
09/05/2006	frmb	added some priority-free calls to CCSP
08/05/2006	frmb	more changes to the run-time interface, CIF updates to use it
02/05/2006	frmb	changed the run-time interface mechanism (mostly)
13/04/2006	frmb	added --svn-user option to the build script
11/04/2006	ats	added command-line option handling to the file library
11/04/2006	ats	small fix for a memory leak in blocking sys-calls handling
10/04/2006	frmb	put 1.4.1-pre3 together
10/04/2006	ats	various updates/fixes for pony
10/04/2006	frmb	added the "cluster DX viewer" to contributed software
10/04/2006	frmb	added Adrian Cockcroft's formatted output library to contributed software
10/04/2006	frmb	put in mechanisms to handle contributed software
10/04/2006	frmb	put 1.4.1-pre2 together
09/04/2006	frmb	added the pony/kroc-net infrastructure to the distribution (experimental)
09/04/2006	frmb	follow-up fix for a related bug
08/04/2006	frmb	fixed a PLACED array bug in occ21, ack to Sergiu Dumitriu for spotting it
06/04/2006	frmb	more interface changes, more multi-processor code
06/04/2006	frmb	fixed a small bug in occ21 (dynamic mobile arrays)
04/04/2006	frmb	changed interface mechanism for some run-time kernel calls
04/04/2006	frmb	fixed a small register colouring problem in tranx86
02/04/2006	frmb	some tidy-ups, more multi-processor code
02/04/2006	frmb	separated out top-level interface processes into libkrocif
02/04/2006	frmb	changed bug-reporting address to kroc-bugs@kent.ac.uk (RT queue)
29/03/2006	frmb	various CCSP tidy-ups, removed old keyboard support
28/03/2006	frmb	started adding some support for multi-processor machines
27/03/2006	frmb	started removing support for scheduler-structures (for a better way)
26/03/2006	frmb	a couple more fixes for pony
25/03/2006	frmb	added some checks in CCSP for invalid mobile processes
25/03/2006	frmb	various updates for pony
22/03/2006	frmb	small fix for compiler crash with missing parameter types, ack to Christian/CO631-students for spotting it
22/03/2006	frmb	fix for a missing check in mobile channel-end communication, ack to Adam for spotting it
21/03/2006	frmb	fixed some more bugs in the usage checker and undefinedness checker
19/03/2006	frmb	fixed a small bug in the undefinedness checker, ack to Adam for finding it
18/03/2006	frmb	put 1.4.1-pre1 together
18/03/2006	frmb	fixed some UDO bugs in the compiler, tidied up other code
18/03/2006	frmb	minor compiler fixes/updates in error paths
18/03/2006	frmb	fixed a type-check bug with anonymous channel-types and protocol inheritance
14/03/2006	frmb	added --linkcc option to kroc wrapper
14/03/2006	frmb	small fix to mixed input/output in arrays of mobile channel-ends passed as parameters
14/03/2006	frmb	small fix to CIF assembler code (.string to .asciz)
13/03/2006	frmb	small fix for -X6 option (extended single-tag inputs)
26/02/2006	frmb	compiler fix disallowing PROTOCOLs to be used as variable types
08/02/2006	frmb	small fix for KRoC.net/pony network CLAIMs
05/02/2006	frmb	various fixes for channel-type type descriptions
01/02/2006	frmb	fixed a small memory-leak in non-freeing of dynamic channel types
22/01/2006	frmb	multiway sync support in the run-time (not for regular occam-pi yet)
14/01/2006	frmb	support for ENROLL instead of BARRIER for doing PAR barrier extension
14/01/2006	frmb	added checks for null mobile barriers on sync/enroll/resign
14/01/2006	frmb	put 1.4.0 together

tools/mkoccdeps/mkoccdeps.c
/*
 *	mkoccdeps.c -- create Makefile dependencies for occam sources
 *	Copyright (C) 2004 Fred Barnes <frmb@kent.ac.uk>
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; either version 2 of the License, or
 *	(at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program; if not, write to the Free Software
 *	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifndef VERSION
#define VERSION "(none)"
#endif

/*{{{  includes*/
#include <stdio.h>

#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#endif

#include <sys/types.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <unistd.h>

#ifdef HAVE_STDARG_H
#include <stdarg.h>
#endif

#include "global.h"
#include "support.h"
/*}}}*/

/*{{{  local and global vars*/
#define FOLLOW_NONE 0
#define FOLLOW_ALL 1
#define FOLLOW_LOCAL 2

static struct {
	int follow;
	int verbose;
	char *outfile;
	int allobjs;
	int localonly;
	DYNARRAY(char *, ipath);
} options = {
	.follow = FOLLOW_NONE,
	.verbose = 0,
	.outfile = NULL,
	.allobjs = 0,
	.localonly = 0
};

#define FTYPE_UNKNOWN 0
#define FTYPE_SOURCE 1
#define FTYPE_INCLUDE 2
#define FTYPE_LIBRARY 3
#define FTYPE_BINARY 4

typedef struct TAG_dfile_t {
	char *name;
	int ftype;
	int skip;
	int hidden;
	int given;
	DYNSARRAY(TAG_dfile_t*, deps);
} dfile_t;

STATICDYNARRAY(dfile_t *, files);


static char *argvzero;
char *progname;
/*}}}*/


/*{{{  void mkoccdeps_fatal (char *fmt, ...)*/
/*
 *	fatal error handler
 */
void mkoccdeps_fatal (char *fmt, ...)
{
	va_list ap;

	va_start (ap, fmt);
	fprintf (stderr, "%s: fatal: ", progname);
	vfprintf (stderr, fmt, ap);
	fprintf (stderr, "\n");
	fflush (stderr);
	va_end (ap);

	exit (EXIT_FAILURE);
}
/*}}}*/
/*{{{  static void usage (FILE *stream)*/
/*
 *	print usage info
 */
static void usage (FILE *stream)
{
	fprintf (stream, "%s %s: occam dependency generator\n", progname, VERSION);
	fprintf (stream, "Copyright (C) 2004 Fred Barnes <frmb@kent.ac.uk>.  GPL >= 2\n\n");
	fprintf (stream, "usage: %s [options] [--] <file1> [file2 [...]]\n", argvzero);
	fprintf (stream, "where options are:\n");
	fprintf (stream, "    -h | --help             print this help and exit\n");
	fprintf (stream, "    -V | --version          print version and exit\n");
	fprintf (stream, "    -v | --verbose          verbose operation\n");
	fprintf (stream, "    -o | --output  <file>   output to <file> (default stdout)\n");
	fprintf (stream, "    -f | --follow           follow referenced files\n");
	fprintf (stream, "    -F | --follow-local     only follow files referenced in the current directory\n");
	fprintf (stream, "    -a | --allobjs          output ALLOBJS entry\n");
	fprintf (stream, "    -A | --allotherobjs     output ALLOBJS entry with all object files\n");
	fprintf (stream, "                            except those given (only makes sense with -f)\n");
	fprintf (stream, "    -I | --include  <path>  search <path> for occam libraries and includes\n");
	fprintf (stream, "    -l | --local-only       only output dependencies for files in the current directory\n");

	return;
}
/*}}}*/
/*{{{  static void version (FILE *stream)*/
/*
 *	print version info
 */
static void version (FILE *stream)
{
	fprintf (stream, "%s %s\n", progname, VERSION);
	return;
}
/*}}}*/

/*{{{  static int guess_ftype (char *fname)*/
/*
 *	guess at file-type based on extension
 */
static int guess_ftype (char *fname)
{
	char *ch;
	int ftype = FTYPE_UNKNOWN;

	if (!fname || !strlen (fname)) {
		mkoccdeps_fatal ("bad file-name");
	}
	for (ch = fname + (strlen (fname) - 1); (ch > fname) && (ch[-1] != '.') && (ch[-1] != '/'); ch--);
	if (ch == fname) {
		ftype = FTYPE_BINARY;
	} else if (!strcmp (ch - 1, ".occ")) {
		ftype = FTYPE_SOURCE;
	} else if (!strcmp (ch - 1, ".inc")) {
		ftype = FTYPE_INCLUDE;
	} else if (!strcmp (ch - 1, ".lib")) {
		ftype = FTYPE_LIBRARY;
	} else {
		ftype = FTYPE_BINARY;
	}

	return ftype;
}
/*}}}*/
/*{{{  static char *find_object_file (char *fname)*/
/*
 *	searches the "include path" for something that produces "fname"
 */
static char *find_object_file (char *fname)
{
	char *tmpname, *ch, *rpath;
	int i;

	if (!fname || !strlen (fname)) {
		return NULL;
	}
	rpath = NULL;

	/* find extension and replace with .occ */
	tmpname = (char *)smalloc (strlen (fname) + 8);
	memcpy (tmpname, fname, strlen (fname) + 1);
	for (ch = tmpname + (strlen (tmpname) - 1); (ch > tmpname) && (*ch != '.'); ch--);
	if (ch == tmpname) {
		sprintf (tmpname, "%s.occ", fname);
		ch = tmpname + strlen (fname);
	} else {
		sprintf (ch, ".occ");
	}

	/* always search here first */
	if (!access (tmpname, R_OK)) {
		/* this source exists here */
		rpath = string_dup (tmpname);
	}
	/* iterate over include directories */
	for (i=0; !rpath && (i < DA_CUR (options.ipath)); i++) {
		char *ipath = DA_NTHITEM (options.ipath, i);

		rpath = (char *)smalloc (strlen (ipath) + strlen (tmpname) + 4);
		if (!strcmp (ipath, "/")) {
			sprintf (rpath, "/%s", tmpname);
		} else {
			sprintf (rpath, "%s/%s", ipath, tmpname);
		}

		if (access (rpath, R_OK)) {
			sfree (rpath);
			rpath = NULL;
		}
	}

	if (!rpath) {
		/* doesn't seem to exist */
		fprintf (stderr, "%s: warning: cannot find source for %s\n", progname, fname);
	} else {
		/* extension must be .occ, set back to .o */
		rpath[strlen (rpath) - 2] = '\0';
	}

	sfree (tmpname);
	return rpath;
}
/*}}}*/
/*{{{  static char *find_include_file (char *fname)*/
/*
 *	searches the "include path" for exactly fname
 */
static char *find_include_file (char *fname)
{
	char *ch, *rpath;
	int i;

	if (!fname || !strlen (fname)) {
		return NULL;
	}
	rpath = NULL;

	/* always search here first */
	if (!access (fname, R_OK)) {
		/* this source exists here */
		rpath = string_dup (fname);
	}
	/* iterate over include directories */
	for (i=0; !rpath && (i < DA_CUR (options.ipath)); i++) {
		char *ipath = DA_NTHITEM (options.ipath, i);

		rpath = (char *)smalloc (strlen (ipath) + strlen (fname) + 4);
		if (!strcmp (ipath, "/")) {
			sprintf (rpath, "/%s", fname);
		} else {
			sprintf (rpath, "%s/%s", ipath, fname);
		}

		if (access (rpath, R_OK)) {
			sfree (rpath);
			rpath = NULL;
		}
	}

	if (!rpath) {
		/* doesn't seem to exist */
		fprintf (stderr, "%s: warning: cannot find %s\n", progname, fname);
	}

	return rpath;
}
/*}}}*/
/*{{{  static int process_file (dfile_t *file)*/
/*
 *	processes the given file.
 *	return 0 on success, -1 on error
 */
static int process_file (dfile_t *file)
{
	if (file->skip) {
		return 0;
	}
	if (options.verbose) {
		fprintf (stderr, "processing %s...\n", file->name);
	}
	if ((file->ftype == FTYPE_SOURCE) || (file->ftype == FTYPE_INCLUDE)) {
		FILE *fp;
		char buf[512];
		int curline = 0;

		if (!(fp = fopen (file->name, "rt"))) {
			/* not a problem if it's an include file */
			if (file->ftype == FTYPE_INCLUDE) {
				file->hidden = 1;
				return 0;
			}
			return -1;
		}
		while (fgets (buf, 511, fp)) {
			char *ch, *dh;
			int include = -1;

			curline++;
			/* skip leading whitespace */
			for (ch = buf; (*ch != '\0') && ((*ch == ' ') || (*ch == '\t')); ch++);
			if (*ch != '#') {
				/* uninteresting */
				continue;
			}
			if (!strncmp (ch, "#USE", 4)) {
				include = 0;
				dh = ch + 4;
			} else if (!strncmp (ch, "#INCLUDE", 8)) {
				include = 1;
				dh = ch + 8;
			} else {
				dh = NULL;
			}
			if (dh) {
				/* attempt to extract string portion */
				char *eh;
				dfile_t *tmp = NULL;
				int ttype;

				for (; (*dh == ' ') || (*dh == '\t'); dh++);
				if (*dh != '\"') {
					/* wierd */
					fprintf (stderr, "%s:%d: badly formed directive\n", file->name, curline);
					continue;
				}
				dh++;

				for (eh=dh; (*eh != '\0') && (*eh != '\"'); eh++);
				if (*eh != '\"') {
					fprintf (stderr, "%s:%s: badly formed directive\n", file->name, curline);
					continue;
				}
				*eh = '\0';
				/* dh now has the USE'd or INCLUDE'd name, create an entry for it */
				ttype = guess_ftype (dh);
				if ((ttype == FTYPE_BINARY) && !include) {
					/* depends on an object-file */
					char *tmpname = (char *)smalloc (strlen (dh) + 8);
					char *tn2;
					dfile_t *tmp2 = NULL;
					int i;
					
					sprintf (tmpname, "%s.o", dh);
					/* search for it */
					tn2 = find_object_file (tmpname);
					if (tn2) {
						sfree (tmpname);
						tmpname = tn2;
					}
					for (i = 0; i < DA_CUR (files); i++) {
						if (!strcmp ((DA_NTHITEM (files, i))->name, tmpname)) {
							break;
						}
					}
					if (i == DA_CUR (files)) {
						/* not here, create an entry */
						tmp = (dfile_t *)smalloc (sizeof (dfile_t));
						tmp->name = string_dup (tmpname);
						tmp->ftype = ttype;
						tmp->skip = 1;				/* no point for an object-file */
						tmp->hidden = 0;
						tmp->given = 0;
						dynarray_init (tmp->deps);
						dynarray_add (files, tmp);
					} else {
						tmp = DA_NTHITEM (files, i);
					}

					/* search for .occ source and add to list */
					sprintf (tmpname, "%s.occ", dh);
					tn2 = find_include_file (tmpname);
					if (tn2) {
						sfree (tmpname);
						tmpname = tn2;
					}
					for (i = 0; i < DA_CUR (files); i++) {
						if (!strcmp ((DA_NTHITEM (files, i))->name, tmpname)) {
							break;
						}
					}
					if (i == DA_CUR (files)) {
						/* not here, create an entry */
						tmp2 = (dfile_t *)smalloc (sizeof (dfile_t));
						tmp2->name = string_dup (tmpname);
						tmp2->ftype = FTYPE_SOURCE;
						tmp2->hidden = 0;
						switch (options.follow) {
						case FOLLOW_NONE:
							tmp2->skip = 1;
							break;
						case FOLLOW_ALL:
							tmp2->skip = 0;
							break;
						case FOLLOW_LOCAL:
							tmp2->skip = ((strchr (tmp2->name, '/')) != NULL);
							break;
						}
						tmp2->given = 0;
						dynarray_init (tmp2->deps);
						dynarray_add (files, tmp2);
					} else {
						tmp2 = DA_NTHITEM (files, i);
					}
					/* object depends on the source */
					dynarray_maybeadd (tmp->deps, tmp2);

					if (tmpname) {
						sfree (tmpname);
						tmpname = NULL;
					}
				} else if ((ttype == FTYPE_INCLUDE) && include) {
					/* depends on include file, search for it */
					int i;
					char *tmpname, *tn2;
					int nonlocal = 0;

					tmpname = string_dup (dh);
					tn2 = find_include_file (tmpname);
					if (tn2) {
						sfree (tmpname);
						tmpname = tn2;
					} else {
						nonlocal = 1;
					}
					for (i = 0; i < DA_CUR (files); i++) {
						if (!strcmp ((DA_NTHITEM (files, i))->name, tmpname)) {
							break;
						}
					}
					if (i == DA_CUR (files)) {
						/* not here, create an entry */
						tmp = (dfile_t *)smalloc (sizeof (dfile_t));
						tmp->name = string_dup (tmpname);
						tmp->ftype = ttype;
						switch (options.follow) {
						case FOLLOW_NONE:
							tmp->skip = 1;
							break;
						case FOLLOW_ALL:
							tmp->skip = 0;
							break;
						case FOLLOW_LOCAL:
							tmp->skip = ((strchr (tmp->name, '/')) != NULL);
							break;
						}
						tmp->hidden = nonlocal;
						tmp->given = 0;
						dynarray_init (tmp->deps);
						dynarray_add (files, tmp);
					} else {
						tmp = DA_NTHITEM (files, i);
					}

					if (tmpname) {
						sfree (tmpname);
						tmpname = NULL;
					}
				}

				/* if tmp is set, link it (if not already a dependency) */
				if (tmp) {
					dynarray_maybeadd (file->deps, tmp);
				}
			}
		}
		fclose (fp);

		if (file->ftype == FTYPE_SOURCE) {
			/* convert into object-file dependencies */
			int i;
			char *tmpname = string_dup (file->name);
			char *ch = tmpname + (strlen (tmpname) - 4);
			dfile_t *tmp;

			if ((ch < tmpname) || strcmp (ch, ".occ")) {
				fprintf (stderr, "%s: warning: strange looking source file: %s\n", progname, file->name);
			} else {
				sprintf (ch, ".o");

				/* search for it */
				for (i = 0; i < DA_CUR (files); i++) {
					if (!strcmp ((DA_NTHITEM (files, i))->name, tmpname)) {
						break;
					}
				}
				if (i == DA_CUR (files)) {
					/* create new entry for object-file */
					tmp = (dfile_t *)smalloc (sizeof (dfile_t));
					tmp->name = tmpname;
					tmp->ftype = FTYPE_BINARY;
					tmp->skip = 1;
					tmp->hidden = 0;
					tmp->given = file->given;
					dynarray_init (tmp->deps);
					dynarray_add (files, tmp);
				} else {
					tmp = DA_NTHITEM (files, i);
				}

				/* move dependencies from file->deps to tmp->deps */
				for (i = 0; i < DA_CUR (file->deps); i++) {
					dfile_t *ditem = DA_NTHITEM (file->deps, i);

					dynarray_maybeadd (tmp->deps, ditem);
				}
				dynarray_trash (file->deps);

				/* and add the new dependency */
				dynarray_maybeadd (tmp->deps, file);
			}
		}
	}
	return 0;
}
/*}}}*/
/*{{{  static void write_deps (dfile_t *file, FILE *stream)*/
/*
 *	outputs dependency info for a file
 */
static void write_deps (dfile_t *file, FILE *stream)
{
	int i;

	if (!file) {
		return;
	}
	if (!DA_CUR (file->deps)) {
		/* no dependencies */
		return;
	}
	if (file->hidden) {
		/* not our problem */
		return;
	}

	if (options.localonly && strchr (file->name, '/')) {
		/* elsewhere */
		return;
	}
	for (i = 0; i < DA_CUR (file->deps); i++) {
		dfile_t *dep = DA_NTHITEM (file->deps, i);

		if (!dep->hidden) {
			fprintf (stream, "%s: %s\n", file->name, dep->name);
		}
	}
	return;
}
/*}}}*/
/*{{{  static void write_allobjs (FILE *stream)*/
/*
 *	outputs an ALLOBJS= ... for Makefiles
 */
static void write_allobjs (FILE *stream)
{
	int i;

	fprintf (stream, "ALLOBJS=");
	for (i = 0; i < DA_CUR (files); i++) {
		dfile_t *file = DA_NTHITEM (files, i);

		if ((file->ftype == FTYPE_BINARY) && !file->hidden) {
			if ((options.allobjs == 2) && !file->given) {
				fprintf (stream, "%s ", file->name);
			} else if (options.allobjs == 1) {
				fprintf (stream, "%s ", file->name);
			}
		}
	}
	fprintf (stream, "\n");
	return;
}
/*}}}*/

/*{{{  int main (int argc, char **argv)*/
/*
 *	start here
 */
int main (int argc, char **argv)
{
	int i;
	char **walk;
	int eoo = 0;

	argvzero = *argv;
	for (progname = *argv + (strlen (*argv) - 1); (progname > *argv) && (progname[-1] != '/'); progname--);
	dynarray_init (files);
	dynarray_init (options.ipath);

	/* process arguments */
	for (walk = argv + 1, i = argc; *walk && i; walk++, i--) {
		if (!eoo && (!strcmp (*walk, "-h") || !strcmp (*walk, "--help"))) {
			usage (stdout);
			exit (EXIT_SUCCESS);
		} else if (!eoo && (!strcmp (*walk, "-V") || !strcmp (*walk, "--version"))) {
			version (stdout);
			exit (EXIT_SUCCESS);
		} else if (!eoo && (!strcmp (*walk, "-v") || !strcmp (*walk, "--verbose"))) {
			options.verbose = 1;
		} else if (!eoo && (!strcmp (*walk, "-f") || !strcmp (*walk, "--follow"))) {
			options.follow = FOLLOW_ALL;
		} else if (!eoo && (!strcmp (*walk, "-F") || !strcmp (*walk, "--follow-local"))) {
			options.follow = FOLLOW_LOCAL;
		} else if (!eoo && (!strcmp (*walk, "-a") || !strcmp (*walk, "--allobjs"))) {
			options.allobjs = 1;
		} else if (!eoo && (!strcmp (*walk, "-A") || !strcmp (*walk, "--allotherobjs"))) {
			options.allobjs = 2;
		} else if (!eoo && (!strcmp (*walk, "-o") || !strcmp (*walk, "--output"))) {
			walk++, i--;
			if (!*walk || !i) {
				mkoccdeps_fatal ("option %s requires an argument", walk[-1]);
			}
			if (!strcmp (*walk, "-")) {
				options.outfile = NULL;		/* use stdout */
			} else {
				options.outfile = string_dup (*walk);
			}
		} else if (!eoo && (!strcmp (*walk, "-l") || !strcmp (*walk, "--local-only"))) {
			options.localonly = 1;
		} else if (!eoo && (!strncmp (*walk, "-I", 2) || !strcmp (*walk, "--include"))) {
			char *opt;

			/* option might be bolted on if -I */
			if (!strncmp (*walk, "-I", 2) && strcmp (*walk, "-I")) {
				opt = string_dup (*walk + 2);
			} else {
				walk++, i--;
				if (!*walk || !i) {
					mkoccdeps_fatal ("option %s requires an argument", walk[-1]);
				}
				opt = string_dup (*walk);
			}
			if ((opt[strlen(opt) - 1] == '/') && (strlen(opt) > 1)) {
				/* remove trailing / */
				opt[strlen(opt) - 1] = '\0';
			}
			dynarray_add (options.ipath, opt);
		} else if (!eoo && !strcmp (*walk, "--")) {
			eoo = 1;
		} else if (!eoo && (**walk == '-')) {
			mkoccdeps_fatal ("unrecognised option: %s", *walk);
		} else {
			/* assume filename */
			int j;

			for (j = 0; j < DA_CUR (files); j++) {
				if (!strcmp ((DA_NTHITEM (files, j))->name, *walk)) {
					/* already got this one */
					break;
				}
			}
			if (j == DA_CUR (files)) {
				struct stat stbuf;
				dfile_t *tmp;

				/* make sure it's a real file first */
				if (stat (*walk, &stbuf) < 0) {
					mkoccdeps_fatal ("unable to stat file: %s", *walk);
				}
				if (!S_ISREG (stbuf.st_mode)) {
					mkoccdeps_fatal ("%s is not a regular file", *walk);
				}
				tmp = (dfile_t *)smalloc (sizeof (dfile_t));
				tmp->name = string_dup (*walk);
				tmp->ftype = guess_ftype (tmp->name);
				tmp->skip = 0;
				tmp->hidden = 0;
				tmp->given = 1;
				dynarray_init (tmp->deps);

				dynarray_add (files, tmp);
			}
		}
	}

	/* check for at least one file */
	if (!DA_CUR (files)) {
		mkoccdeps_fatal ("no input files");
	}

	/* process */
	for (i = 0; i < DA_CUR (files); i++) {
		dfile_t *fitem = DA_NTHITEM (files, i);

		if (process_file (fitem) < 0) {
			mkoccdeps_fatal ("error processing %s", fitem->name);
		}
	}

	/* output */
	{
		FILE *outfile;

		if (!options.outfile) {
			outfile = stdout;
		} else {
			outfile = fopen (options.outfile, "wt");
			if (!outfile) {
				mkoccdeps_fatal ("unable to open %s for output", options.outfile);
			}
		}
		for (i = 0; i < DA_CUR (files); i++) {
			dfile_t *fitem = DA_NTHITEM (files, i);

			write_deps (fitem, outfile);
		}
		if (options.allobjs) {
			write_allobjs (outfile);
		}
		if (options.outfile) {
			fclose (outfile);
		}
	}

	return 0;
}
/*}}}*/

